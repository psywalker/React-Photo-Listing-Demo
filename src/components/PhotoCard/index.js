import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import {
  Tag,
  Popover,
} from 'antd';
import InfiniteScroll from 'react-infinite-scroller';
import Tags from '../Tags';
import './index.scss';

export const getTagsVisibleArr = item => (
  item.tags.length < 4
    ? item.tags
    : item.tags.slice(0, 3)
);

export const getTagsHiddenArr = item => (
  item.tags.length > 3
    ? item.tags.slice(3)
    : false
);

const PhotoCard = memo(({
  onSearchTagValue,
  getPaginationChange,
  cards,
}) => (
  <InfiniteScroll
    pageStart={0}
    element="ul"
    className="photo-card-list"
    data-test="photoCardList"
    initialLoad={false}
    threshold={50}
    loadMore={getPaginationChange}
    loader={<li className="loader" key={0}>Loading ...</li>}
    hasMore
  >
    { cards.map((item) => {
      const tagsVisibleArr = getTagsVisibleArr(item);
      const tagsHiddenArr = getTagsHiddenArr(item);
      return (
        <li
          data-test="photoCardListItem"
          key={item.photoID}
          className="photo-card-list__item"
        >
          <div
            data-test="photoCard"
            className="photo-card"
          >
            <Link
              data-test="photoCardPhotoLink"
              to={`/photo/${item.photoID}`}
            >
              <img
                data-test="photoCardImg"
                className="photo-card__img"
                alt={item.photoAltDesc}
                src={item.photoName}
              />
            </Link>

            <div
              data-test="photoCardAutor"
              className="photo-card__autor photo-card-autor"
            >
              <Link
                data-test="photoCardAutorLink"
                className="photo-card-autor__link"
                to={`/users/${item.userID}`}
              >
                <img
                  data-test="photoCardAutorAvatar"
                  className="photo-card-autor__avatar"
                  alt={item.userID}
                  src={item.userAvatar}
                />
                <span
                  data-test="photoCardAutorName"
                  className="photo-card-autor__name"
                >
                  { item.title }
                </span>
              </Link>
            </div>

            <p
              className="photo-card__desc"
              data-test="photoCardDesc"
            >
              { item.photoDesc }
            </p>

            <div
              className="photo-card__badge-container photo-card-badge"
              data-test="photoCardBadge"
            >
              <Tags
                data-test="photoCardTagMainContainer"
                handleMethod={onSearchTagValue}
                tags={tagsVisibleArr}
              />
              { tagsHiddenArr && (
                <Popover
                  data-test="photoCardPopover"
                  placement="top"
                  title="Remaining tags"
                  content={(
                    <Tags
                      data-test="photoCardTagPopup"
                      handleMethod={onSearchTagValue}
                      tags={tagsHiddenArr}
                    />
                  )}
                  trigger="click"
                >
                  <Tag
                    data-test="photoCardTagMore"
                    className="photo-card-badge__tag"
                  >
                    more tags...
                  </Tag>
                </Popover>
              )}
            </div>
          </div>
        </li>
      );
    })}
  </InfiniteScroll>
));


PhotoCard.propTypes = {
  onSearchTagValue: PropTypes.func,
  getPaginationChange: PropTypes.func,
  cards: PropTypes.arrayOf(PropTypes.object),
};
PhotoCard.defaultProps = {
  onSearchTagValue: () => {},
  getPaginationChange: () => {},
  cards: [],
};

export default PhotoCard;
