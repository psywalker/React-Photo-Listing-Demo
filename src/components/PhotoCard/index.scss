.photo-card-list {
  margin: 0;
  padding: 0;
  list-style: none;
  margin-left: -10px;
  display: flex;

  &__item {
    position: relative;
    width: calc(100%/3 - 10px);
    margin: 0 0 10px 10px;
    @media screen and ( min-width: 741px ) and ( max-width: 980px ) {
      width: calc(100%/2 - 10px);
    }
    @media screen and ( min-width: 0px ) and ( max-width: 740px) {
      width: calc(100% - 10px);
    }

    &:hover {
      & .photo-card__desc-container {
       
      }
    }
  }
}

.photo-card {
  &__img {
    display: block;
    width: 100%;
  }
  &__desc-container {
    position: absolute;
    bottom: 0;
    left: 0;
    width: 100%;
    opacity: 0;
    transition: all .3s;
  }
  &__desc {
    display: none;
    text-align: center;
    padding: 0 15px;
    text-overflow: ellipsis;
    overflow: hidden;
    white-space: nowrap;
  }
}

.photo-card-autor {
  position: absolute;
  left: 16px;
  top: 5px;
  &__link {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 50px;
  }

  &__avatar {
    width: 30px;
    height: 30px;
    border-radius: 100%;
    margin-right: 16px;
  }

  &__name {
    overflow: hidden;
    color: #fff;
    text-shadow: 0 1px 8px rgba(0,0,0,.1);
    opacity: .8;
    font-weight: 500;
    font-size: 18px;
    white-space: nowrap;
    text-overflow: ellipsis;

    transition: all .3s;
  }

  &:hover .photo-card-autor__name {
    color: #fff;
    opacity: 1;
  }
}

.photo-card-badge {
  position: absolute;
  left: 16px;
  bottom: 16px;
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  align-items: center;

  //display: none;

  // &__tag {
  //   margin: 5px !important;
  //   cursor: pointer  !important;
  // }
  & .tag {
    cursor: pointer;
    opacity: .8;

    &:hover {
      opacity: 1;
    }
  }
}