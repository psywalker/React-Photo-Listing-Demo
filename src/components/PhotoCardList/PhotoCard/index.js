import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { useSelector } from 'react-redux';
import PhotoCardImg from '../PhotoCardImg';
import PhotoCardTags from '../PhotoCardTags';
import PhotoCardAutor from '../PhotoCardAutor';
import LikesButton from '../../LikesButton';
import DownLoadButton from '../../DownLoadButton';
import './index.scss';

const PhotoCard = memo(({
  onSearchTagValue,
  item,
}) => {
  const login = useSelector(state => state.login);
  const { profileName } = login;

  return (
    <li className="photo-card-list-item photo-card-list__item">
      <header className="photo-card-list__header">
        <PhotoCardAutor
          userID={item.userID}
          userAvatar={item.userAvatar}
          title={item.title}
        />

        { profileName && <LikesButton photoID={item.photoID} /> }
      </header>

      <PhotoCardImg
        photoID={item.photoID}
        photoAltDesc={item.photoAltDesc}
        photoName={item.photoName}
      />

      <footer className="photo-card-list__footer">
        <PhotoCardTags
          onSearchTagValue={onSearchTagValue}
          tags={item.tags}
        />

        <DownLoadButton
          altDescriprion={item.photoAltDesc}
          photoDesc={item.photoDesc}
          photoSrc={item.photoName}
          photoUrlSizes={item.photoUrlSizes}
          placement="leftBottom"
          userID={item.userID}
          title={item.title}
        />
      </footer>
    </li>
  );
});


PhotoCard.propTypes = {
  onSearchTagValue: PropTypes.func,
  item: PropTypes.shape({}),
};
PhotoCard.defaultProps = {
  onSearchTagValue: () => {},
  item: {},
};

export default PhotoCard;
