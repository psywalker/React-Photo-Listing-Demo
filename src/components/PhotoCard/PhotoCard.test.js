import React from 'react';
import PhotoCard from '.';
import { getTagsHiddenArr } from '.';

describe('Test of component of `PhotoCard`', () => {
  const initialProps = {
    onSearchTagValue: () => {},
    cards: [],
  };

  const cardsOneElement = [
    {
      photoDesc: null,
      photoAltDesc: 'image',
      photoID: '1',
      tags: [
        {
          title: 'motorcycle',
        },
      ],
      photoName: 'https://images.unsplash.com/photo-1558981408-db0ecd8a1ee4?ixlib=rb-1.2.1&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=1080&fit=max&ixid=eyJhcHBfaWQiOjQzODA4fQ',
      title: 'Harley-Davidson',
      userAvatar: 'https://images.unsplash.com/profile-1556751276456-1561737ea797?ixlib=rb-1.2.1&q=80&fm=jpg&crop=faces&cs=tinysrgb&fit=crop&h=128&w=128',
      userID: 'harleydavidson',
    },
  ];
  const photoCardList = 'ul[data-test="photoCardList"]';
  const photoCardListItem = 'li[data-test="photoCardListItem"]';
  const photoCard = 'div[data-test="photoCard"]';
  const photoCardPhotoLink = 'a[data-test="photoCardPhotoLink"]';
  const photoCardPhotoLinkRouter = 'Link[data-test="photoCardPhotoLink"]';
  const photoCardImg = 'img[data-test="photoCardImg"]';
  const photoCardAutor = 'div[data-test="photoCardAutor"]';
  const photoCardAutorLink = 'a[data-test="photoCardAutorLink"]';
  const photoCardAutorLinkRouter = 'Link[data-test="photoCardAutorLink"]';
  const photoCardAutorAvatar = 'img[data-test="photoCardAutorAvatar"]';
  const photoCardAutorName = 'span[data-test="photoCardAutorName"]';
  const photoCardDesc = 'p[data-test="photoCardDesc"]';
  const photoCardBadge = 'div[data-test="photoCardBadge"]';
  const photoCardTagMainContainer = '[data-test="photoCardTagMainContainer"]';
  const tag = 'div[data-test="tag"]';
  const photoCardTagMore = 'div[data-test="photoCardTagMore"]';

  const appSelector = wrapper => ({
    getPhotoCardList: () => wrapper.find(photoCardList),
    getPhotoCardListItem: () => wrapper.find(photoCardListItem),
    getNthPhotoCardListItem: n => wrapper.find(photoCardListItem).at(n),
    getPhotoCardPhotoLink: () => wrapper.find(photoCardPhotoLink),
    getNthPhotoCardPhotoLink: n => wrapper.find(photoCardPhotoLink).at(n),
    getPhotoCardPhotoLinkRouter: () => wrapper.find(photoCardPhotoLinkRouter),
    getPhotoCard: () => wrapper.find(photoCard),
    getPhotoCardImg: () => wrapper.find(photoCardImg),
    getNthPhotoCardImg: n => wrapper.find(photoCardImg).at(n),
    getPhotoCardAutor: () => wrapper.find(photoCardAutor),
    getPhotoCardAutorLink: () => wrapper.find(photoCardAutorLink),
    getPhotoCardAutorLinkRouter: () => wrapper.find(photoCardAutorLinkRouter),
    getNthPhotoCardAutorLink: n => wrapper.find(photoCardAutorLink).at(n),
    getPhotoCardAutorAvatar: () => wrapper.find(photoCardAutorAvatar),
    getPhotoCardAutorName: () => wrapper.find(photoCardAutorName),
    getPhotoCardDesc: () => wrapper.find(photoCardDesc),
    getPhotoCardBadge: () => wrapper.find(photoCardBadge),
    getNthPhotoCardBadge: n => wrapper.find(photoCardBadge).at(n),
    getTag: () => wrapper.find(tag),
    getNthTag: n => wrapper.find(tag).at(n),
    getPhotoCardTagMainContainer: () => wrapper.find(photoCardTagMainContainer),
    getPhotoCardTagMore: () => wrapper.find(photoCardTagMore),
  });

  describe('`PhotoCard` initional render', () => {
    it('renders without initial props', () => {
      const photoCardComponent = global.mountWrap(<PhotoCard />);
      const page = appSelector(photoCardComponent);
      const cardList = page.getPhotoCardList();
      const cardListItem = page.getPhotoCardListItem();

      expect(cardList).toHaveLength(1);
      expect(cardListItem).toHaveLength(0);

      expect(photoCardComponent).toMatchSnapshot();
    });
    it('renders with initial props', () => {
      const photoCardComponent = global.mountWrap(<PhotoCard {...initialProps} />);
      const page = appSelector(photoCardComponent);
      const cardList = page.getPhotoCardList();
      const cardListItem = page.getPhotoCardListItem();

      expect(cardList).toHaveLength(1);
      expect(cardListItem).toHaveLength(0);

      expect(photoCardComponent).toMatchSnapshot();
    });
    it('renders with one element in `cards`', () => {
      const props = {
        ...initialProps,
        cards: cardsOneElement,
      };
      const photoCardComponent = global.mountWrap(<PhotoCard {...props} />);
      const page = appSelector(photoCardComponent);

      const cardList = page.getPhotoCardList();
      const cardListItem = page.getPhotoCardListItem();

      expect(cardList).toHaveLength(1);
      expect(cardListItem).toHaveLength(1);
      expect(cardListItem).toExist();
      expect(cardList).toContainMatchingElement('.photo-card-list__item');

      expect(photoCardComponent).toMatchSnapshot();
    });
  });

  describe('Test `getTagsHiddenArr`', () => {
    it('result `getTagsHiddenArr` with `< 4` tags to equal `false`', () => {
      const obj = {
        tags: [
          {
            title: 'motorcycle',
          },
          {
            title: 'auto',
          },
          {
            title: 'velocity',
          },
        ],
      };
      const resultTag = getTagsHiddenArr(obj);
      expect(resultTag).toEqual(false);
    });
    it('result `getTagsHiddenArr` with `> 3` tags to equal `false`', () => {
      const obj = {
        tags: [
          {
            title: 'motorcycle',
          },
          {
            title: 'auto',
          },
          {
            title: 'velocity',
          },
          {
            title: 'bike',
          },
        ],
      };
      const resultTag = getTagsHiddenArr(obj);
      expect(resultTag.length).toEqual(1);
    });
  });
  describe('Test `photoCardPhotoLink`', () => {
    it('Test `photoCardPhotoLink` address', () => {
      const props = {
        ...initialProps,
        cards: cardsOneElement,
      };
      const photoCardComponent = global.mountWrap(<PhotoCard {...props} />);
      const page = appSelector(photoCardComponent);
      const cardPhotoLink0 = page.getNthPhotoCardPhotoLink(0);
      const cardPhotoLinkRouter = page.getPhotoCardPhotoLinkRouter(); 
      
      expect(cardPhotoLink0.props().href).toEqual('/photo/1');
      expect(cardPhotoLinkRouter.prop('to')).toEqual('/photo/1');

      expect(photoCardComponent).toMatchSnapshot();
    });
    it('Test data `cards` in component', () => {
      const props = {
        ...initialProps,
        cards: cardsOneElement,
      };
      const photoCardComponent = global.mountWrap(<PhotoCard {...props} />);
      const page = appSelector(photoCardComponent);
      const cardImg = page.getPhotoCardImg();
      const cardAutorLink = page.getPhotoCardAutorLink();
      const cardAutorLinkRouter = page.getPhotoCardAutorLinkRouter();
      const cardAutorAvatar = page.getPhotoCardAutorAvatar();
      const cardAutorName = page.getPhotoCardAutorName();
      const cardPhotoDesc = page.getPhotoCardDesc();
      const cardTagMainContainer = page.getPhotoCardTagMainContainer();

      expect(cardImg.prop('alt')).toEqual('image');
      expect(cardImg.prop('src')).toEqual(cardsOneElement[0].photoName);
      expect(cardAutorLink.props().href).toEqual('/users/harleydavidson');
      expect(cardAutorLinkRouter.prop('to')).toEqual('/users/harleydavidson');
      expect(cardAutorAvatar.prop('alt')).toEqual('harleydavidson');
      expect(cardAutorAvatar.prop('src')).toEqual(cardsOneElement[0].userAvatar);
      expect(cardAutorName).toHaveText('Harley-Davidson');
      expect(cardPhotoDesc).toHaveText('');
      expect(cardTagMainContainer.prop('tags')).toBeArray();
      expect(cardTagMainContainer.prop('handleMethod')).toBeFunction();

      expect(photoCardComponent).toMatchSnapshot();
    });
  });

  describe('Test Tags', () => {
    it('render `tags` on one element `cards`', () => {
      const props = {
        ...initialProps,
        cards: [
          {
            ...cardsOneElement[0],
            tags: [
              {
                title: 'motorcycle',
              },
              {
                title: 'auto',
              },
              {
                title: 'velocity',
              },
              {
                title: 'electric bike',
              },
              {
                title: 'harley-davidson',
              },
              {
                title: 'scooter',
              },
            ],
          },
        ],
      };
      const photoCardComponent = global.mountWrap(<PhotoCard {...props} />);
      const page = appSelector(photoCardComponent);
      const cardPhotoCardTagMore = page.getPhotoCardTagMore();
      const cardTag = page.getTag();
      const cardTag0 = page.getNthTag(0);

      expect(cardTag).toHaveLength(3);
      expect(cardTag0).toHaveText('motorcycle');
      expect(cardPhotoCardTagMore).toHaveText('more tags...');
      expect(cardPhotoCardTagMore).toHaveLength(1);

      expect(photoCardComponent).toMatchSnapshot();
    });
    it('render `tags` on two element `cards`', () => {
      const props = {
        ...initialProps,
        cards: [
          {
            ...cardsOneElement[0],
            tags: [
              {
                title: 'motorcycle1',
              },
              {
                title: 'auto1',
              },
              {
                title: 'velocity1',
              },
            ],
          },
          {
            ...cardsOneElement[0],
            photoID: '2',
            tags: [
              {
                title: 'motorcycle2',
              },
              {
                title: 'auto2',
              },
              {
                title: 'velocity2',
              },
            ],
          },
        ],
      };
      const photoCardComponent = global.mountWrap(<PhotoCard {...props} />);
      const page = appSelector(photoCardComponent);
      const cardTag = page.getTag();
      const cardTag5 = page.getNthTag(5);
      
      expect(cardTag).toHaveLength(6);
      expect(cardTag5).toHaveText('velocity2');

      expect(photoCardComponent).toMatchSnapshot();
    });

    it('Mock `onSearchTagValue` ', () => {
      const mockFetchRequestAction = jest.fn();
      const props = {
        cards: [
          {
            ...cardsOneElement[0],
            tags: [
              {
                title: 'motorcycle1',
              },
              {
                title: 'auto1',
              },
              {
                title: 'velocity1',
              },
              {
                title: 'motorcycle2',
              },
              {
                title: 'auto2',
              },
              {
                title: 'velocity2',
              },
            ],
          },
        ],
        onSearchTagValue: mockFetchRequestAction,
      };

      const photoCardComponent = global.mountWrap(<PhotoCard {...props} />);
      const page = appSelector(photoCardComponent);
      const cardBadge = page.getPhotoCardBadge();
      const cardTag = page.getTag();
      const cardTag0 = page.getNthTag(0);
      cardTag0.simulate('click');
      expect(mockFetchRequestAction).toHaveBeenCalledTimes(1);
    });
  });
});