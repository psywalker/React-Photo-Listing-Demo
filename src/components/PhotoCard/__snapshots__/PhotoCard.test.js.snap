// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of component of PhotoCard Home component initial renders with \`cards\` 1`] = `
<ul
  className="photo-list"
  data-test="PhotoList"
>
  <li
    className="photo-list__item"
    data-test="PhotoListItem"
    key="kdGstD3te3M"
  >
    <div
      className="photo-card-self"
    >
      <Card
        cover={
          <Link
            className="photo-card__photo-link"
            replace={false}
            to="/photo/kdGstD3te3M"
          >
            <img
              alt="example"
              className="photo-card__img"
              src="https://images.unsplash.com/photo-1558981408-db0ecd8a1ee4?ixlib=rb-1.2.1&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=1080&fit=max&ixid=eyJhcHBfaWQiOjQzODA4fQ"
            />
          </Link>
        }
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <Link
          className="photo-card-self__link-ava"
          replace={false}
          to="/users/harleydavidson"
        >
          <Meta
            avatar={
              <Avatar
                shape="circle"
                size="default"
                src="https://images.unsplash.com/profile-1556751276456-1561737ea797?ixlib=rb-1.2.1&q=80&fm=jpg&crop=faces&cs=tinysrgb&fit=crop&h=128&w=128"
              />
            }
            title="Harley-Davidson"
          />
        </Link>
        <p
          className="photo-card-self__desc"
        >
          No Description
        </p>
        <div
          className="photo-card-self__badge-wrap"
        >
          <Tag
            className="photo-card-self__badge"
            closable={false}
            key="motorcycle"
            onClick={[Function]}
          >
            motorcycle
          </Tag>
        </div>
      </Card>
    </div>
  </li>
</ul>
`;

exports[`Test of component of PhotoCard Home component initial renders with initial props 1`] = `
<ul
  className="photo-list"
  data-test="PhotoList"
/>
`;
